Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift

```diff
@@ -152,6 +168,26 @@ enum SyntaxStringInterpolationError: Error, CustomStringConvertible {
   }
 }
 
+/// A Swift type whose value can be represented directly in source code by a Swift literal.
+///
+/// Conforming types do not *contain* Swift source code; rather, they can be *expressed* in Swift source code, and this protocol can be used to get whatever source code would do that. For example, `String` is `ExpressibleByLiteralSyntax` but `StringLiteralExprSyntax` is not.
+///
+/// Conforming types can be interpolated into a Swift source code literal with the syntax `\(literal: <value>)`:
+///
+///      let greeting = "Hello, world!"
+///      let expr1 = ExprSyntax("print(\(literal: greeting))")
+///      // `expr1` is a syntax tree for `print("Hello, world!")`
+///
+/// Note that quote marks are automatically added around the contents; you don't have to write them yourself. The conformance will automatically ensure the contents are correctly escaped, possibly by using raw literals or other language features:
+///
+///      let msPath = "c:\\windows\\system32"
+///      let expr2 = ExprSyntax("open(\(literal: msPath))")
+///      // `expr2` might be a syntax tree for `open(#"c:\windows\system32"#)`
+///      // or for `open("c:\\windows\\system32")`.
+public protocol ExpressibleByLiteralSyntax {
+  func makeLiteralSyntax() -> ExprSyntaxProtocol
```

Comment:
Should we introduce an associated type here to express the result type more completely, e.g.,

```swift
  associatedtype LiteralSyntaxNode: ExprSyntaxProtocol
  func makeLiteralSyntax() -> LiteralSyntaxNode
```

so that (e.g.) `String`'s `makeLiteralSyntax` produces a `StringLiteralExprSyntax`?

Reply:
For floats, I ended up needing to return some things that weren't strictly literal expressions, like `-.infinity`. Would I just say that floats map to `ExprSyntax` or something?

Reply:
I haven't done this because of both floats and the new `Optional` conformance (which sometimes generates `.some(nil)` and such), but we can discuss it if you think it's important.